lib/puppet/configurer.rb

```diff
@@ -386,6 +386,15 @@ def run_internal(options)
         query_options, facts = get_facts(options)
         query_options[:configured_environment] = configured_environment
 
+        new_env = Puppet::Node::Environment.remote(@environment)
+        Puppet.push_context(
+          {
+            :current_environment => new_env,
+            :loaders => Puppet::Pops::Loaders.new(new_env, true)
+          },
+          "Local node environment #{@environment} for configurer transaction"
+        )
```

Comment:
Should this be moved outside of the `while loop`? to push the context only once, after the correct environment is found?

Reply:
Also, we execute this piece of code 3 times in this file, would it be worth it to extract it to a separate method?